### PR TITLE
156 donor search Step function step

### LINKED ISSUE #156: Donor search process
We need to find donors with the following parameters. 

1. Filter by the last donation date. (User specified)
2. location filter. (User specified)
3. Blood group filter. (User specified)
4. Filter by vaccination. (Backend)
5. Opt-out fitter for travel or any other unspecified reason. (Backend)
6. Include blood bank and organization in search results. (Backend and option in app end to skip)
7. Skip incomplete doners. (Backend)

Backend marked parameters should be handled behind the scenes for app users and User specified will be included input for org-users. Check [here](https://drive.google.com/file/d/1ya9swGl5iBPYcSVBcL0NmUqBUIE5DHvI/view?usp=sharing) for detailed design

**Proposed Solution:**
This will be part of the donor search and notification step specified in the detailed design. This also includes tuning geohash length and retry dynamodb search with tuned geohash length. As we are searching donor details it can easily exceeds step function's data limit, we can use step function's distributed map to tackle the issue

##Task breakdown : 

- [ ] prepare dynamodb search parameter in query generation step with retry logic the adjust with dynamodb RCU. Query generation step must include selective fields that we need for notification.
- [ ] Dynamodb integration and store results in s3 for later usage

## Acceptance Criteria
- [ ] Elegable donors are found from the dynamodb and stored in s3 for later process

## Documentation References
- Check [here](https://drive.google.com/file/d/1ya9swGl5iBPYcSVBcL0NmUqBUIE5DHvI/view?usp=sharing) for detailed design

### PR DESCRIPTION
**Pull request by mahinuzzaman**
Fixes #156 
- Added donor search process for stepfunction
- updated sone existing functions and unittests

Resolves #
Used when the changes in the pull request indirectly resolve another issue.

- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [x] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

Provide a brief overview of the changes introduced in this pull request.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [ ] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [ ] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

File: iac/terraform/aws/donor_search/stepFunction.tf [hunk 1]
```diff
@@ -73,6 +80,7 @@ resource "aws_sfn_state_machine" "donor_search_state_machine" {
 
   definition = templatefile("${path.module}/donor_search.json", {
     DONOR_CALCULATE_LAMBDA_ARN_PLACEHOLDER = module.step_function_lambda["calculate-donors-to-notify"].lambda_arn
```
Comment (melon-ruet) [inline iac/terraform/aws/donor_search/stepFunction.tf:82]: Don't use placeholder in param name

File: iac/terraform/aws/donor_search/stepFunction.tf [hunk 2]
```diff
@@ -73,6 +80,7 @@ resource "aws_sfn_state_machine" "donor_search_state_machine" {
...
+    GEOHASH_CALCULATE_LAMBDA_ARN_PLACEHOLDER = module.step_function_lambda["donor-search-evaluator"].lambda_arn
```
Comment (melon-ruet) [inline iac/terraform/aws/donor_search/stepFunction.tf:83]: Don't use placeholder in param name

Comment (mahinuzzaman) [reply iac/terraform/aws/donor_search/stepFunction.tf]: updated in 0d5f0245edf91b83b3cc1ea2f7c82e86225e14dc

File: core/services/aws/bloodDonation/donorSearchEvaluator.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+interface DonorSearchInput {
```
Comment (sakibstark11) [inline core/services/aws/bloodDonation/donorSearchEvaluator.ts:1]: any reason why this is interface

**Review by sakibstark11 — APPROVED** [review]

File: core/services/aws/bloodDonation/donorSearchEvaluator.ts (hunk 3 above)
Comment (mahinuzzaman) [reply core/services/aws/bloodDonation/donorSearchEvaluator.ts:1]: vai, in all lambda we are using an interface to take input

**Review by melon-ruet — APPROVED** [review]

**Pull request merged**